pkg/api/ref.go

```diff
@@ -52,23 +52,41 @@ func GetReference(obj runtime.Object) (*ObjectReference, error) {
 	if err != nil {
 		return nil, err
 	}
-	_, kind, err := Scheme.ObjectVersionAndKind(obj)
-	if err != nil {
-		return nil, err
+
+	// if the object referenced is actually persisted, we can just get kind from meta
+	// if we are building an object reference to something not yet persisted, we should fallback to scheme
+	kind := meta.Kind()
+	if kind == "" {
+		_, kind, err = Scheme.ObjectVersionAndKind(obj)
+		if err != nil {
+			return nil, err
+		}
 	}
-	version := ""
-	parsedSelfLink := versionFromSelfLink.FindStringSubmatch(meta.SelfLink())
-	if len(parsedSelfLink) < 2 {
-		if ForTesting_ReferencesAllowBlankSelfLinks {
-			version = "testing"
-		} else if meta.SelfLink() == "" {
-			return nil, ErrNoSelfLink
+
+	// if the object referenced is actually persisted, we can also get version from meta
+	version := meta.APIVersion()
+	if version == "" {
```

Comment:
Sorry for extremely late delay, but getting the version from the self link should be first in the priority order; otherwise this will end up lying about where the object came from.